### DISCUSSION TITLE
Keyboard vs UI Jogging

### DISCUSSION THREAD

**Discussion by phil-barrett** [terjeio/ioSender#181] [General]
This has been gnawing at me for a while.  I will admit, I have not fully understood jogging in all its' complexities until now (and perhaps my understanding is still incomplete). I think the distinction between UI and keyboard jogging is very confusing. UI jogging causes people to see a jog rate/distance that keyboard jogging doesn't adhere to. And, there is nothing in the app to point this out. The differences between keyboard jogging and UI jogging are not immediately obvious. Honestly, I would make the keyboard arrow keys behave the same as the UI keys. One jogging panel to show all the information related to jogging. I do realize the keyboard jogging rates/distances are available/visible if the underlying grblHAL has the keyboard enabled and you go to settings. Though, that is not at all obvious to a typical user.

Jogging input can come from devices that are directly connect to the grblHAL motion controller.  This adds a lot of really great possibilities but adds complexity.  In particular, it drives the need to have some jogging control info in grbl settings.

I've looked at other GCode Senders.  UGS is very simplistic, just having keyboard step functionality.  Their UI jog buttons behave the same as keyboard buttons. Open Builds Control is more sophisticated - it has a visible switch between what they call Incremental and Continuous Jogging.  Roughly analogous to UI and Keyboard Jogging in ioSender. The on screen ("UI") buttons and keyboard arrow keys behave the same in both cases though they also suffer from the Windows UI interception issue. There is one slider - Jog - that controls jogging.  Switching to Incremental shows a distance increment selection (.1, 1, 10, 100).  The Jog slider controls the feed rate. It goes from 1% to 100%, repeat is not supported. I dislike the unitless approach and would prefer a set of selected rates.  In Continuous Jogging, the distance selection turns off. The Jog slider controls feed rate - it moves as long as the key is down so faster jog rate = farther distance for a given length key press.  While I dislike the OB approach, at least they make it clear which jogging mode is active and it is relatively simple.  Watching new CNC users with OB Control, they picked up both styles of jogging pretty quickly.  ioSender was harder for them to understand - they seldom strayed from the UI controls.

So, what would I do if I could wave a magic wand? Well, I would move everything related to jogging into the the jogging panel. Ensure consistency between keyboard keys and UI buttons. I would use a switch to select between the two kinds of jogging - step vs repeat (and make it persistent across ioSender invocations). Make it clear what the settings are. Ideally, you could change the grblHAL jog settings from the jog panel (though, I can understand wanting to keep grbl settings management in one place). Also, what ever mode is in use should be clear from the jog panel (if non-UI changing is allowed).  Keyboard jogging with 3 settings (step/slow/fast) is ok to me but not being able to use it all the time is a huge confusion point. Maybe allow non-arrow keys for jogging to avoid clashing with Windows UI and allow game controllers and PC connected pendants/mpgs to always work. Maybe ctrl+shift+[udlrhl] as a set of alternative motion control keystrokes. I could also see using something similar for directly setting jog rates/distances. It would be really great if the keys were configurable.

**Comment by terjeio** [comment]
> Well, I would move everything related to jogging into the the jogging panel.

Ok, I will make it a setting that requires editing the config file to switch to keyboard jogging as it is now. When enabled this will remove the jog panel from the UI as well to reduce confusion. There should be no action on key up in the new mode?

**Reply by terjeio** [reply]
> There should be no action on key up in the new mode?

Jogging will be started on key up for now, press _End_ to stop jogging.

**Reply by terjeio** [reply]
> Maybe allow non-arrow keys for jogging to avoid clashing with Windows UI and allow game controllers and PC connected pendants/mpgs to always work. Maybe ctrl+shift+[udlrhl] as a set of alternative motion control keystrokes.

IIRC older games had the cursor key alternatives clustered in nearby keys - that would be better than having them spread all over the keyboard? I have to admit that I have not played any games since the mid eighties so I really do not know...

> I could also see using something similar for directly setting jog rates/distances.

Added in edge build 36p24 by using numpad 0-3 for step size and numpad 4-7 for distance (with Ctrl modifier).

> It would be really great if the keys were configurable.

Maybe in build 37, the keypress handler is centralized but not key to function mapping. Controls tells the keypress handler which keypresses they want to subscribe to, I guess somehow controls has to publish (to a mapping class) which functions are available for making this possible - I guess via a viewmodel and corresponding UI.

**Reply by calabr** [reply]
Probable you need middleware event dispetcher, which will subscribe to keyboard events and call internal events according to keyboard map

**Reply by phil-barrett** [reply]
Thank you and Merry Christmas!

**Reply by phil-barrett** [reply]
> IIRC older games had the cursor key alternatives clustered in nearby keys - that would be better than having them spread all over the keyboard? I have to admit that I have not played any games since the mid eighties so I really do not know...

The actual key choices aren't that important since the most likely use is via gamepads, dongles and MPGs.

As I recall a few Unix games from the dark ages uses hjkl for x and y movement - on a qwerty kbd for easy use via one hand without moving the hand.

**Reply by terjeio** [reply]
> Probable you need middleware event dispetcher, which will subscribe to keyboard events and call internal events according to keyboard map

I have a dispatcher but need code/UI for handling the map, controls having shortcuts register them directly via a method call to the dispatcher. They have to register them with a map handler instead which in turn will register them with the dispatcher based on user input.

**Reply by terjeio** [reply]
> As I recall a few Unix games from the dark ages uses hjkl for x and y movement - on a qwerty kbd for easy use via one hand without moving the hand.

Ok, it will be like this in 36p25 then (with shift+ctrl modifiers):

`private const Key xplus = Key.H, xminus = Key.J, yplus = Key.K, yminus = Key.L, zplus = Key.I, zminus = Key.M;`

**Reply by calabr** [reply]
Is it simple to follow Unix way? Locate mapping in config file and read it on startup.
 Setup window and change on the fly can wait.

**Reply by terjeio** [reply]
> Is it simple to follow Unix way? Locate mapping in config file and read it on startup.

Yes, after code is written to parse it and available methods are registered with the parser. The parser has to save available method names and current keybindings to a file so it can be edited by the user - but I guess when the parser is ready it will be simple (and likely safer) to add a UI on top of it. Next version...

**Reply by terjeio** [reply]
@calabr 
> Is it simple to follow Unix way? Locate mapping in config file and read it on startup.
Setup window and change on the fly can wait.

Available now in [latest edge versions](https://github.com/terjeio/ioSender/discussions/176#discussioncomment-2261131).

**Reply by cuishuang413** [reply]
> > As I recall a few Unix games from the dark ages uses hjkl for x and y movement - on a qwerty kbd for easy use via one hand without moving the hand.
> 
> Ok, it will be like this in 36p25 then (with shift+ctrl modifiers):
> 
> `private const Key xplus = Key.H, xminus = Key.J, yplus = Key.K, yminus = Key.L, zplus = Key.I, zminus = Key.M;`

How would you change speed and distance with shift and ctrl exactly? What are "modifiers"? I have tried pressing shift or/and ctrl and arrow keys, page up and down keys, and all other keys on the keyboard, but nothing changes. I always have to use the mouse to select on the UI. Am I missing something?

Thank you!

**Reply by terjeio** [reply]
@cuishuang413 You have to create a [shortcut to ioSender.exe](https://github.com/terjeio/ioSender/wiki/Setup-and-configuration#optional-command-line-parameters) and add `-jogmode 1` (or 2) to enable keyboard jogging.

**Reply by cuishuang413** [reply]
Got it. Thanks a lot!
